pkg/aws/sts.go

```diff
@@ -202,3 +204,99 @@ func BuildOperatorRoleCommands(prefix string, accountID string, awsClient Client
 	}
 	return commands
 }
+
+func (c *awsClient) FindAccountRoleArnBundles(version string) (map[string]map[string]string, error) {
+
+	roles, err := c.ListRoles()
+	if err != nil {
+		return map[string]map[string]string{}, err
+	}
+	prefixesCountMap := make(map[string]int)
+	prefixesARNMap := make(map[string]map[string]string)
+	for _, role := range roles {
+		foundPrefix, foundType, ok := isMatchWithAccountRolePrefix(role)
+		if !ok {
+			continue
+		}
+		if _, mapOk := prefixesCountMap[foundPrefix]; !mapOk {
+			prefixesCountMap[foundPrefix] = 0
+			prefixesARNMap[foundPrefix] = make(map[string]string)
+		}
+		prefixesCountMap[foundPrefix]++
+
+		isTagged, skip, err := c.checkRoleTagsCompatibility(role, version)
+		if err != nil {
+			return map[string]map[string]string{}, err
+		}
+		if isTagged && !skip {
+			prefixesCountMap[foundPrefix]++
+			prefixesARNMap[foundPrefix][foundType] = *role.Arn
+		}
+	}
+
+	return extractCompleteAccountBundles(prefixesCountMap, prefixesARNMap), nil
+}
+
+func (c *awsClient) checkRoleTagsCompatibility(role *iam.Role, version string) (bool, bool, error) {
+	listRoleTagsOutput, err := c.iamClient.ListRoleTags(&iam.ListRoleTagsInput{
+		RoleName: role.RoleName,
+	})
+	if err != nil {
+		return false, false, err
+	}
+	skip := false
+	isTagged := false
+	for _, tag := range listRoleTagsOutput.Tags {
+		tagValue := aws.StringValue(tag.Value)
+		switch aws.StringValue(tag.Key) {
+		case tags.RoleType:
+			isTagged = true
+		case tags.OpenShiftVersion:
+			isTagged = true
+			clusterVersion, err := semver.NewVersion(version)
+			if err != nil {
+				skip = true
+				break
+			}
+			policyVersion, err := semver.NewVersion(tagValue)
+			if err != nil {
+				skip = true
+				break
+			}
+			if policyVersion.LessThan(clusterVersion) {
+				skip = true
+				break
+			}
+		}
+	}
+	return isTagged, skip, nil
+}
+
+func isMatchWithAccountRolePrefix(role *iam.Role) (foundPrefix string, foundType string, ok bool) {
+	matches := PrefixAccRoleRE.FindStringSubmatch(*role.RoleName)
+	if len(matches) == 0 {
+		return "", "", false
+	}
+	prefixIndex := PrefixAccRoleRE.SubexpIndex("Prefix")
+	typeIndex := PrefixAccRoleRE.SubexpIndex("Type")
+	foundPrefix = strings.ToLower(matches[prefixIndex])
+	foundType = matches[typeIndex]
+	return foundPrefix, foundType, true
+
+}
+
+func extractCompleteAccountBundles(prefixesCountMap map[string]int,
+	prefixesArnMap map[string]map[string]string) map[string]map[string]string {
+	bundles := make(map[string]map[string]string)
+	for key, value := range prefixesCountMap {
+		isBundleComplete := value == len(AccountRoles)*2
```

Comment:
Why do we have to multiply by two?
A complete bundle contains four account role

Reply:
The first checks if it finds the prefix 4 times one for each type of role. The second is that it checks the compatibility with the version supplied and fetches the actual arn. It is a hacky solution to only using one map. It can definitely be refactored. If you think it is worth it I'll change it as soon as I get through some other higher priorities

Reply:
Ok, it's your call.
Maybe it is worth adding a comment about the calculation, just for other team members in the future.